Library/Homebrew/formula_auditor.rb

```diff
@@ -332,6 +332,23 @@ def audit_deps
       end
 
       return unless @core_tap
+
+      # Prevent formulae from declaring Linux-only dependencies on GCC,
+      # or else its dependency tree grows out of control.
+      on_linux_deps = formula.to_hash_with_variations.dig("variations", :x86_64_linux, "dependencies")
```

Comment:
I don't think this will work when running on Linux. `to_hash_with_variations` creates the variation hash with respect to whatever OS is currently being used. So if we run this on Linux, there will never be an `:x86_64_linux` key in the variations hash, but there might be one for every version of macOS.

For example, here's a formula with a linux-only gcc dep:

```ruby
on_linux do
  depends_on "gcc"
end
```

On macOS, `#to_hash_with_variations` will include:

```ruby
{
  ...
  "dependencies" => [],
  ...
  "variations" => {
    x86_64_linux: { dependencies: ["gcc"] },
  },
}
```

On Linux, `#to_hash_with_variations` will include:

```ruby
{
  ...
  "dependencies" => ["gcc"],
  ...
  "variations" => {
    ventura: { dependencies: [] },
    monterey: { dependencies: [] },
    arm64_ventura: { dependencies: [] },
    arm64_monterey: { dependencies: [] },
    ...
  },
}
```

Reply:
Ah, that's unexpected. Will tweak this a bit then. Thanks!

Reply:
> `to_hash_with_variations` creates the variation hash with respect to whatever OS is currently being used.

Is this with respect to the "real" OS, or the simulated OS? Do we ever make a distinction?

Reply:
Well technically simulated, but with a catch. Whatever OS we're currently on (whether simulated or real) when the formula is _first loaded_ is what will be used. Because of the way formulae are loaded, we have to do some [extra work](https://github.com/Homebrew/brew/blob/b556db72fba0ccd8edf361c6e5e7b4036abad9dc/Library/Homebrew/formula.rb#L2057-L2061) to load the alternate versions without conflicts.

It's a little extra work, but you could probably do something like this if you wanted (not tested):

```ruby
def linux_only_deps
  formula_contents = formula.path.read
  macos_formula, linux_formula = [:macos, :linux].each do |os|
    Homebrew::SimulateSystem.os = os
    Homebrew::SimulateSystem.arch = :intel if os == :linux

    variations_namespace = Formulary.class_s("Variations#{os.capitalize}")
    variations_formula_class = Formulary.load_formula(formula.name, formula.path, formula_contents, variations_namespace,
                                                      flags: formula.class.build_flags, ignore_errors: true)
    variations_formula = variations_formula_class.new(formula.name, formula.path, :stable,
                                                      alias_path: formula.alias_path, force_bottle: formula.force_bottle)

    Homebrew::SimulateSystem.clear
    variations_formula
  end

  linux_formula.deps.map(&:name) - macos_formula.deps.map(&:name)
end
```

Reply:
Could we maybe just do this audit on Linux only? Feels like that'd be a lot simpler.

Reply:
The tap syntax job sets `HOMEBREW_SIMULATE_MACOS_ON_LINUX` so this should be fine to run only on macOS (as long as we use `Homebrew::SimulateSystem.simulating_or_running_on_macos?` instead of `OS.mac?` to check which we're running on.

Reply:
I guess the downside of that is that contributors using Linux won't see the `brew audit` failure. This is particularly concerning when adding a new formula that might need a Linux-only GCC dependency. A bit tempted to punt on this one, though.